src/features/map/MapVisualizer.tsx

```diff
@@ -55,45 +70,143 @@ function getStaticMapOptions(){
   };
 }
 
-export default function MapVisualizer({ map, interactable = false }: Props): JSX.Element {
-  const memoizedCreateGraph = useMemo(() => createGraph(map), [map]);
+export default function MapVisualizer({
+  map,
+  interactable = false,
+  select,
+  doubleClick,
+  onNodeMoved,
+  edgeCreation,
+}: Props): JSX.Element {
+  const [mapState, setMapState] = useState(createGraph(map));
   const additionalOptions = interactable ? {} : getStaticMapOptions();
+  const edgeCreationMode = edgeCreation?.value;
+  const setEdgeCreationMode = edgeCreation?.setter;
+
+  function handleDoubleClick(event: any) {
+    if (doubleClick) {
+      const newNodeData = doubleClick(
+        Math.round(event.pointer.canvas.x / DISTANCE_MULTIPLIER),
+        Math.round(event.pointer.canvas.y / DISTANCE_MULTIPLIER),
+      );
+
+      setMapState((graph: GraphData) => {
+        return {
+          nodes: [
+            ...graph.nodes,
+            createNode(
+              random(100, 999999),
+              newNodeData.name,
+              newNodeData.type,
+              newNodeData.position,
+            ),
+          ],
+          edges: [...graph.edges],
+        };
+      });
+    }
+  }
+
+  function handleDragEnd(event: any) {
+    if (event.nodes.length === 1) {
+      const nodeMovedId = event.nodes[0];
+      const nodeMoved = mapState.nodes.find((n) => n.id === nodeMovedId);
+      if (nodeMoved && onNodeMoved) {
+        // using position of mouse pointer, so some drags can be little bit off,
+        // but I couldn't find exact position where node had moved ;(
+        const newX = Math.round(event.pointer.canvas.x / DISTANCE_MULTIPLIER);
+        const newY = Math.round(event.pointer.canvas.y / DISTANCE_MULTIPLIER);
+
+        nodeMoved.x = newX * DISTANCE_MULTIPLIER;
+        nodeMoved.y = newY * DISTANCE_MULTIPLIER;
+
+        onNodeMoved(newX, newY, nodeMoved.label ?? '');
+      }
+    }
+  }
+
+  function handleSelectInternal(clickedNode: Node) {
+    const createEdgeToSelf = edgeCreationMode?.firstNodeName == clickedNode.label;
+    const createEdge = edgeCreationMode && edgeCreationMode.modeOn && edgeCreationMode.firstNodeName;
+    if (createEdgeToSelf) {
+      return;
+    }
+    if (createEdge) {
+      const firstNode = mapState.nodes.find((n) => n.label === edgeCreationMode.firstNodeName);
+      if (firstNode) {
+        setMapState((graph: GraphData) => {
+          return {
+            nodes: graph.nodes,
+            edges: [
+              { from: firstNode.id, to: clickedNode.id, value: 1 },
+              ...graph.edges,
+            ],
+          };
+        });
+      }
+    }
+  }
+
+  function handleSelect(nodes: any) {
+    if (nodes.length == 1) {
+      const nodeId = nodes[0];
+      const clickedNode = mapState.nodes.find((n) => n.id === nodeId);
+      if (clickedNode && clickedNode.label && select) {
+        select(clickedNode.label);
```

Comment:
czemu po labelu akurat? to api biblioteczki?

Reply:
Label to nazwa i właśnie tu jest lekkie ryzyko bo po nazwie znajduje node w edytorze, bo uzupełniając dane w edytorze nie wpisujemy id nody i przez to nie da się po id rozróżnić obiektu

Reply:
XDDDD no dobra